File: 3383-taking-maximum-energy-from-the-mystic-dungeon/3383-taking-maximum-energy-from-the-mystic-dungeon.cpp
class Solution {
public:
    int maximumEnergy(vector<int>& energy, int k) {
        int res = INT_MIN;
        vector<int> dp;
        dp.push_back(energy[0]);
        for(int i = 1; i < energy.size(); i++)
        {
            if(i - k >= 0)
            {
                dp.push_back(max(energy[i], dp[i - k] + energy[i]));
            }
            else
            {
                dp.push_back(energy[i]);
            }
            if(energy.size() - i <= k)
            {
                res = max(res, dp[i]);
            }
        }
        return res;
    }
};

File: 3383-taking-maximum-energy-from-the-mystic-dungeon/3383-taking-maximum-energy-from-the-mystic-dungeon.py
class Solution:
    def maximumEnergy(self, energy: List[int], k: int) -> int:
        res = float('-inf')
        dp = []
        dp.append(energy[0])
        for i in range(1, len(energy)):
            if i - k >= 0:
                dp.append(max(energy[i], dp[i - k] + energy[i]))
            else:
                dp.append(energy[i])
            if len(energy) - i <= k:
                res = max(res, dp[i])
        return res